optimizer/resolver.go

```diff
@@ -0,0 +1,568 @@
+// and puts the table source in current resolverContext.
+func (nr *nameResolver) handleTableSource(ts *ast.TableSource) {
+	for _, v := range ts.GetResultFields() {
+		v.TableAsName = ts.AsName
+	}
+	var name string
+	if ts.AsName.L != "" {
+		name = ts.AsName.L
+	} else {
+		tableName := ts.Source.(*ast.TableName)
+		name = nr.tableUniqueName(tableName.Schema, tableName.Name)
+	}
+	ctx := nr.currentContext()
+	if _, ok := ctx.tableMap[name]; ok {
+		nr.Err = errors.Errorf("duplicated table/alias name %s", name)
+		return
+	}
+	ctx.tableMap[name] = len(ctx.tables)
+	ctx.tables = append(ctx.tables, ts)
+	return
+}
+
+// handleJoin sets result fields for join.
+func (nr *nameResolver) handleJoin(j *ast.Join) {
+	if j.Right == nil {
+		j.SetResultFields(j.Left.GetResultFields())
+		return
+	}
+	leftLen := len(j.Left.GetResultFields())
+	rightLen := len(j.Right.GetResultFields())
+	rfs := make([]*ast.ResultField, leftLen+rightLen)
+	copy(rfs, j.Left.GetResultFields())
+	copy(rfs[leftLen:], j.Right.GetResultFields())
+	j.SetResultFields(rfs)
+}
+
+// handleColumnName looks up and sets ResultField for
+// the column name.
+func (nr *nameResolver) handleColumnName(cn *ast.ColumnNameExpr) {
+	ctx := nr.currentContext()
+	if ctx.inOnCondition {
+		// In on condition, only tables within current join is available.
+		nr.resolveColumnNameInOnCondition(cn)
+		return
+	}
+
+	// Try to resolve the column name form top to bottom in the context stack.
+	for i := len(nr.contextStack) - 1; i >= 0; i-- {
+		if nr.resolveColumnNameInContext(nr.contextStack[i], cn) {
+			// Column is already bound or encountered an error.
+			return
+		}
+	}
+	nr.Err = errors.Errorf("unknown column %s", cn.Name.Name.L)
+}
+
+// resolveColumnNameInContext looks up and sets ResultField for a column with the ctx.
+func (nr *nameResolver) resolveColumnNameInContext(ctx *resolverContext, cn *ast.ColumnNameExpr) (done bool) {
+	if cn.Name.Table.L == "" {
+		// If qualified table name is not specified in column name, the column name may be ambiguous,
+		// We need to iterate over all tables and
+	}
+
+	if ctx.inTableRefs {
+		// In TableRefsClause, column reference only in join on condition which is handled before.
+		return false
+	}
+	if ctx.inFieldList {
+		// only resolve column using tables.
+		return nr.resolveColumnInTableSources(cn, ctx.tables)
+	}
+	if ctx.inGroupBy {
+		// From tables first, then field list.
+		if ctx.inByItemExpression {
+			// From table first, then field list.
+			if nr.resolveColumnInTableSources(cn, ctx.tables) {
+				return true
+			}
+			return nr.resolveColumnInResultFields(cn, ctx.fieldList)
+		}
+		// Field list first, then from table.
+		if nr.resolveColumnInResultFields(cn, ctx.fieldList) {
+			if nr.Err != nil {
+				return true
+			}
+			// The column is resolved in field list, but we need to
+			// try to overwrite the ResultField in table sources.
+			nr.resolveColumnInTableSources(cn, ctx.tables)
+			if nr.Err != nil {
+				nr.Err = nil
```

Comment:
It is resolved already, we just try to resolve it in table sources, any error should be ignore.